Valkey 8.0 RC1 Blog (#105)
valkey-io/valkey-io.github.io: content/blog/2024-07-31-valkey-8-0-0-rc1.md

# For the most part, you can leave this as the day you _started_ the post.
# The maintainers will update this value before publishing
# The time is generally irrelvant in how Valkey published, so '01:01:01' is a good placeholder
date= 2024-08-02 01:01:01
# 'description' is what is shown as a snippet/summary in various contexts.
# You can make this the first few lines of the post or (better) a hook for readers.
# Aim for 2 short sentences.
description= "Delivering significant improvements in throughput, scalability, and observability for modern data-intensive applications"
# 'authors' are the folks who wrote or contributed to the post.
# Each author corresponds to a biography file (more info later in this document)
authors= [ "pingxie", "madolson" ]
+++

# Valkey 8.0: Unleashing Unprecedented Performance and Reliability

The Valkey team is proud to unveil the first release candidate of Valkey 8.0,
Valkey 8.0 RC1, a major update designed to enhance performance, reliability,
and observability for both standalone setups and clusters of various sizes.

## Performance

Valkey 8.0 features a significant improvement to the existing I/O threading system, allowing the main thread and

> I/O threads to operate concurrently. This release also introduces a number of improvements to offload or work to the I/O threads and introduces efficient batching of commands. All told, Valkey 8 is designed to up to 1.2 million Queries Per Second (QPS) on AWS's r7g

Reviewer: ```suggestion
I/O threads to operate concurrently. 
This release also introduces a number of improvements to offload more work to the I/O threads and introduces batching of commands to improve the efficiency of executing commands on modern hardware.
All told, Valkey 8 is designed to execute up to 1.2 million Queries Per Second (QPS) on AWS's r7g
```

the replica backlog to be transferred simultaneously, accelerating synchronization.

- **Reduced Memory Load**: By streaming replication data to the replica during
  the full sync, the primary node experiences significantly less memory pressure.
  The replica now manages the Client Output Buffer (COB) tracking, reducing the
  likelihood of COB overruns and enabling larger COB sizes on the replica side.
- **Lighter Main Process Load**: A dedicated connection for RDB transfer frees
  the primary's main process from handling this data, allowing it to focus on
  client queries and improving overall responsiveness.

Performance tests show improvements in write latency during sync, and in scenarios
with heavy read commands, the sync time can be cut by up to 50%. This translates
to a more responsive system, even during synchronization.

## Observability

Valkey 8.0 introduces a comprehensive per-slot metrics infrastructure, providing
detailed visibility into the performance and resource usage of individual slots.
This granular data helps inform decisions about resource allocation, load
balancing, and performance optimization.

> - **CPU Usage**: Tracks CPU time consumed by operations on each slot, identifying

Reviewer: ```suggestion
- **Key Count**: Returns the number of keys in each slot, making it easier to identify the slots with the largest number of keys.
- **CPU Usage**: Tracks CPU time consumed by operations on each slot, identifying
```

key pointers and significantly reducing memory overhead. This results in a 9-10%
reduction in overall memory usage for scenarios with 16-byte keys and 8 or 16-byte
values, along with slight performance improvements.

## Additional Highlights

- **Dual IPv4 and IPv6 Stack Support**: Seamlessly operate in mixed IP environments
  for enhanced compatibility and flexibility.
- **Improved Pub/Sub Efficiency**: Lightweight cluster messages streamline
  communication and reduce overhead for faster, more efficient Pub/Sub operations.
- **Valkey Over RDMA (Experimental)**: Unlock significant performance improvements
  with direct memory access between clients and Valkey servers, delivering up to
  275% increase in throughput.
- **Numerous Smaller Performance/Reliability Enhancements**: Many under-the-hood
  improvements ensure a smoother, more stable experience across the board.

## Conclusion

Valkey 8.0 is a significant update that offers improved performance, reliability,
and observability. Whether you are an experienced Valkey/Redis user or exploring
it for the first time, this release provides significant advancements in in-memory

> data storage. Upgrade today to experience these enhancements.

Reviewer: ```suggestion
data storage.
You can try out these enhancements today by downloading from [source](https://github.com/valkey-io/valkey/releases/tag/8.0.0-rc1) or using one of our [container images](https://hub.docker.com/r/valkey/valkey).
```

+++
# `title` is how your post will be listed and what will appear at the top of the post
title= "Valkey 8.0: Enhancing Performance and Reliability"
# `date` is when your post will be published.
# For the most part, you can leave this as the day you _started_ the post.
# The maintainers will update this value before publishing
# The time is generally irrelvant in how Valkey published, so '01:01:01' is a good placeholder
date= 2024-07-31 01:01:01
# 'description' is what is shown as a snippet/summary in various contexts.
# You can make this the first few lines of the post or (better) a hook for readers.
# Aim for 2 short sentences.
description= "Delivering significant improvements in throughput, scalability, and observability for modern data-intensive applications"
# 'authors' are the folks who wrote or contributed to the post.
# Each author corresponds to a biography file (more info later in this document)
authors= [ "pingxie" ]
+++

> # Valkey 8.0: Unleashing Unprecedented Performance and Reliability

Reviewer 2: Suggestion - Delivering Enhanced Performance and Reliability

> # Valkey 8.0: Unleashing Unprecedented Performance and Reliability

Reviewer: ```suggestion
# Valkey 8.0: Unleashing Enhanced Performance and Reliability
```

Author: going with @katiemcollins' edit

The Valkey team is proud to unveil the first release candidate of Valkey 8.0,
Valkey 8.0 RC1, a major update designed to enhance performance, reliability,

> and observability for both standalone setups and clusters of various sizes.

Reviewer: ```suggestion
and observability for both all installations of Valkey.
```

> 

Reviewer 2: in the intro, do we want to say anything about the collaboration across TSC? could be a good opportunity to highlight partnerships

Author: It is actually really the community's collective efforts. I will change to "The Valkey community" instead

## Performance

> Valkey 8.0 features asynchronous I/O threading, allowing the main thread and
> I/O threads to operate concurrently. This enhancement significantly increases
> throughput, achieving up to 1.2 million Queries Per Second (QPS) on AWS's r7g
> platform, compared to the previous limit of 380K QPS.

Reviewer: ```suggestion
Valkey 8.0 features a significant improvement to the existing I/O threading system, allowing the main thread and
I/O threads to operate concurrently. This release also introduces a number of improvements to offload or work to the I/O threads and introduces efficient batching of commands. All told, Valkey 8 is designed to up to 1.2 million Queries Per Second (QPS) on AWS's r7g
platform, compared to the previous limit of 380K QPS. 

NOTE: All improvements are not yet available in the release candidate, but will be available for the GA of Valkey 8.0.
```

> Valkey 8.0 focuses on intelligent optimization, not just raw speed. By fine-
> tuning CPU core utilization, minimizing command latency, and incorporating
> Memory Access Amortization (MAA) to streamline dictionary operations, Valkey
> 8.0 delivers an in-memory data store that is both efficient and responsive.

Reviewer: ```suggestion
```
blah blah blah. I don't think this paragraph says anything. I also really still don't like Memory Access Amortization (MAA). It's just prefetching.

- **Asynchronous I/O Threading**: Enables parallel processing of commands and
  I/O operations, maximizing throughput and minimizing bottlenecks.

> - **Optimized CPU Core Utilization**: Distributes I/O tasks across multiple
>   cores, reducing idle time and maximizing computational power.

Reviewer: ```suggestion
- **Intelligent Core Utilization**: Distributes I/O tasks across multiple
  cores based on realtime usage, reducing idle time and improving energy efficiency.
```

> - **Memory Access Amortization (MAA)**: Optimizes memory access patterns by
>   prefetching frequently accessed data to minimize CPU cache misses, reducing
>   memory accesses required for dictionary operations.

Reviewer: ```suggestion
- **Command batching**: Optimizes memory access patterns by
  prefetching frequently accessed data to minimize CPU cache misses, reducing
  memory accesses required for dictionary operations.
```

Reviewer: I refuse to call this Memory Access amortization.

## Reliability

> Cluster scaling operations, particularly re-sharding and slot migrations, have

Reviewer: Reposting earlier comment. What is the different between resharding and slot migration? I use those interchangeably.

Author: I consider "re-sharding" as the end result and "slot migration" as the mechanism. but reading this again, I see the confusion too. will rephrase

historically been delicate. Valkey 8.0 improves reliability and minimizes
disruptions with the following enhancements:

- **Automatic Failover for Empty Shards**: New shards that start empty, owning
  no slots, now benefit from automatic failover. This ensures high availability
  from the start of the scaling process.
- **Replication of Slot Migration States**: All `CLUSTER SETSLOT` commands are

>   now replicated synchronously on replicas before execution on the primary. This

Reviewer: ```suggestion
  now replicated synchronously to replicas before execution on the primary. This
```

>   guarantees a consistent slot migration state across the shard, preventing data
>   loss or inconsistencies if the primary fails.

Reviewer: ```suggestion
  reduces the chance of unavailability in the event the primary fails, as the replicas has the most up to date information about the state of the shard. 
    New Replicas also automatically inherit the state from primary, with no additional input needed from an operator.
```
You really want to guarantee stuff I don't think we can guarantee :)

Author: Perfect. I like your version. I am too proud of my work... proximity bias

- **Logical Link Recovery**: In the event of a failover, Valkey 8.0 automatically
  repairs the logical links between source and target nodes. This ensures requests
  are correctly routed to the new primary in the target shard, maintaining cluster

>   integrity and availability.

Reviewer: This point still confuses, I think saying there is a "logical link" between source and target is confusing because it implies a TCP link or a clusterbus link. We have `cluster links` for example. Maybe **Slot Migration State recovery** instead?

> - **Seamless Support for New Replicas**: New replicas joining a shard mid-migration
>   will automatically inherit the correct slot migration state, ensuring immediate
>   consistency.

Reviewer: ```suggestion
```
I still think this is a nuance of the previous points. I suggested an edit.

## Replication

Valkey 8.0 introduces a dual-channel replication scheme, allowing the RDB and
the replica backlog to be transferred simultaneously, accelerating synchronization.

- **Reduced Memory Load**: By streaming replication data to the replica during
  the full sync, the primary node experiences significantly less memory pressure.
  The replica now manages the Client Output Buffer (COB) tracking, reducing the
  likelihood of COB overruns and enabling larger COB sizes on the replica side.

> - **Lighter Main Process Load**: A dedicated connection for RDB transfer frees
>   the primary's main process from handling this data, allowing it to focus on
>   client queries and improving overall responsiveness.

Reviewer: ```suggestion
- **Reduced Parent Process Load**: A dedicated connection for RDB transfer frees
  the primary's parent process from handling this data, allowing it to focus on
  client queries and improving overall responsiveness.
```
Typically parent is used instead of main process. But, not sure people will necessarily understand what this means. I almost just wonder if we should say "Reduced Primary load".

Performance tests show improvements in write latency during sync, and in scenarios
with heavy read commands, the sync time can be cut by up to 50%. This translates

> to a more responsive system, even during synchronization.

Reviewer: I don't think we should put numbers without a source. Can we link the PR here?

## Observability

Valkey 8.0 introduces a comprehensive per-slot metrics infrastructure, providing
detailed visibility into the performance and resource usage of individual slots.
This granular data helps inform decisions about resource allocation, load
balancing, and performance optimization.

> - **CPU Usage**: Tracks CPU time consumed by operations on each slot, identifying

Reviewer: Missing key count.

  areas of high utilization and potential bottlenecks.
- **Network Input/Output Bytes**: Monitors data transmission and reception by
  each slot, offering insights into network load and bandwidth utilization.
- **Minimal Overhead**: Initial benchmarks show that enabling detailed metrics
  incurs a negligible overhead of approximately 0.7% in QPS.

## Efficiency

Valkey 8.0 introduces key embedding in the main dictionary, eliminating separate
key pointers and significantly reducing memory overhead. This results in a 9-10%
reduction in overall memory usage for scenarios with 16-byte keys and 8 or 16-byte

> values, along with slight performance improvements.

Reviewer: Also per-slot dictionary. I feel like we discussed this.

## Additional Highlights

- **Dual IPv4 and IPv6 Stack Support**: Seamlessly operate in mixed IP environments
  for enhanced compatibility and flexibility.
- **Improved Pub/Sub Efficiency**: Lightweight cluster messages streamline
  communication and reduce overhead for faster, more efficient Pub/Sub operations.
- **Valkey Over RDMA (Experimental)**: Unlock significant performance improvements
  with direct memory access between clients and Valkey servers, delivering up to

>   275% increase in throughput and over 50% latency reduction in certain scenarios.

Reviewer: ```suggestion
  275% increase in throughput.
```
My personal take is we aren't materially reducing latency here, so we shouldn't call it out. Latency is a very sensitive topic, and normally the result of queueing. Throughput is pretty easy to measure.

- **Numerous Smaller Performance/Reliability Enhancements**: Many under-the-hood
  improvements ensure a smoother, more stable experience across the board.

## Conclusion

> Valkey 8.0 is a comprehensive update that offers improved performance, reliability,

Reviewer: ```suggestion
Valkey 8.0 is a significant update that offers improved performance, reliability,
```
Or major. Just don't say comprehensive, since that is vague.

> and functionality. Whether you are an experienced Valkey/Redis user or exploring

Reviewer: ```suggestion
and observability. Whether you are an experienced Valkey/Redis user or exploring
```
Just to be consistent with the description. Not much new functionality here.

it for the first time, this release provides significant advancements in in-memory

> data storage. Upgrade today to experience these enhancements.

Reviewer: Yeah. Let me propose an edit for that.

# You can make this the first few lines of the post or (better) a hook for readers.
# Aim for 2 short sentences.
description= "Delivering significant improvements in throughput, scalability, and observability for modern data-intensive applications"
# 'authors' are the folks who wrote or contributed to the post.
# Each author corresponds to a biography file (more info later in this document)
authors= [ "pingxie", "madolson" ]
+++

# Valkey 8.0: Delivering Enhanced Performance and Reliability

The Valkey community is proud to unveil the first release candidate of Valkey 8.0,
Valkey 8.0 RC1, a major update designed to enhance performance, reliability,
and observability for all Valkey installations.

## Performance

Valkey 8.0 features significant improvements to the existing I/O threading system,
allowing the main thread and I/O threads to operate concurrently. This release also
includes a number of improvements to offload work to the I/O threads and introduces
efficient batching of commands. Altogether, Valkey 8.0 is designed to handle up to
1.2 million Queries Per Second (QPS) on AWS's r7g platform, compared to the previous

> limit of 380K QPS. 

Reviewer: ```suggestion
limit of 380K QPS. We'll dive deeper into these numbers in an upcoming blog.
```

  an operator.
- **Logical Link Recovery**: In the event of a failover, Valkey 8.0 automatically
  repairs the logical links between source and target nodes. This ensures requests
  are correctly routed to the new primary in the target shard, maintaining cluster
  integrity and availability.

## Replication

Valkey 8.0 introduces a dual-channel replication scheme, allowing the RDB and
the replica backlog to be transferred simultaneously, accelerating synchronization.

- **Reduced Memory Load**: By streaming replication data to the replica during
  the full sync, the primary node experiences significantly less memory pressure.
  The replica now manages the Client Output Buffer (COB) tracking, reducing the
  likelihood of COB overruns and enabling larger COB sizes on the replica side.
- **Reduced Parent Process Load**: A dedicated connection for RDB transfer frees
  the primary's parent process from handling this data, allowing it to focus on
  client queries and improving overall responsiveness.

Performance tests show improvements in write latency during sync, and in scenarios
with heavy read commands, the sync time can be cut by up to 50%. This translates

> to a more responsive system, even during synchronization.

Reviewer: ```suggestion
to a more responsive system, even during synchronization. See [#60](https://github.com/valkey-io/valkey/pull/60) for details.
```

## Observability

Valkey 8.0 introduces a comprehensive per-slot metrics infrastructure, providing
detailed visibility into the performance and resource usage of individual slots.
This granular data helps inform decisions about resource allocation, load
balancing, and performance optimization.

- **Key Count**: Returns the number of keys in each slot, making it easier to identify the slots with the largest number of keys.
- **CPU Usage**: Tracks CPU time consumed by operations on each slot, identifying
  areas of high utilization and potential bottlenecks.
- **Network Input/Output Bytes**: Monitors data transmission and reception by
  each slot, offering insights into network load and bandwidth utilization.
- **Minimal Overhead**: Initial benchmarks show that enabling detailed metrics
  incurs a negligible overhead of approximately 0.7% in QPS.

## Efficiency

> Valkey 8.0 introduces key embedding in the main dictionary, eliminating separate

Reviewer: ```suggestion
Valkey 8.0 introduces two new improvements that reduce the memory overhead of keys, allowing users to store more data without any application changes.
The change change is that keys are now embedded in the main dictionary, eliminating separate
```
This section felt like it should have more of an introduction, similar to the other sections.

key pointers and significantly reducing memory overhead. This results in a 9-10%
reduction in overall memory usage for scenarios with 16-byte keys and 8 or 16-byte

> values, along with performance improvements.

Reviewer: ```suggestion
values, along with performance improvements. 

Valkey 8.0 also introduces a new per-slot dictionary for Valkey cluster, which replaces a linked list that used to allow operator to list out all the keys in a slot for slot-migration.
The new architecture splits the main dictionary by slot, reducing the memory overhead by 16 bytes per key-value pair without degrading performance. 
See [#541](http://github.com/valkey-io/valkey/pulls/541) for more detail.
```

## Additional Highlights

- **Dual IPv4 and IPv6 Stack Support**: Seamlessly operate in mixed IP environments
  for enhanced compatibility and flexibility.
- **Improved Pub/Sub Efficiency**: Lightweight cluster messages streamline
  communication and reduce overhead for faster, more efficient Pub/Sub operations.
- **Valkey Over RDMA (Experimental)**: Unlock significant performance improvements
  with direct memory access between clients and Valkey servers, delivering up to

>   275% increase in throughput.

Reviewer: ```suggestion
  275% increase in throughput. See [#477](https://github.com/valkey-io/valkey/pull/477) for details.
```

- **Numerous Smaller Performance/Reliability Enhancements**: Many under-the-hood
  improvements ensure a smoother, more stable experience across the board.

## Conclusion

> Valkey 8.0 is a major update that offers improved performance, reliability, and
> observability. Whether you are an experienced Valkey/Redis user or exploring
> it for the first time, this release provides significant advancements in in-memory
> data storage. You can try out these enhancements today by downloading from [source]
> (https://github.com/valkey-io/valkey/releases/tag/8.0.0-rc1) or using one of our
> [container images](https://hub.docker.com/r/valkey/valkey). We would love to hear
> your thoughts on these new features and what you hope to see in the future from
> the Valkey project.

Reviewer: ```suggestion
Valkey 8.0 is a major update that offers improved performance, reliability, and
observability. Whether you are an experienced Valkey/Redis user or exploring
it for the first time, this release provides significant advancements in in-memory
data storage. You can try out these enhancements today by downloading from 
[source](https://github.com/valkey-io/valkey/releases/tag/8.0.0-rc1) or using one 
of our [container images](https://hub.docker.com/r/valkey/valkey). We would love 
to hear your thoughts on these new features and what you hope to see in the 
future from the Valkey project.
```

> You can try out these enhancements today by downloading from [source](https://github.com/valkey-io/valkey/releases/tag/8.0.0-rc1) or using one of our [container images](https://hub.docker.com/r/valkey/valkey).

Reviewer: ```suggestion
You can try out these enhancements today by downloading from [source](https://github.com/valkey-io/valkey/releases/tag/8.0.0-rc1) or using one of our [container images](https://hub.docker.com/r/valkey/valkey).
We would love to hear your thoughts about these new features and what you hope to see in the future from the Valkey project.
```

+++
# `title` is how your post will be listed and what will appear at the top of the post
title= "Valkey 8.0: Delivering Enhanced Performance and Reliability"
# `date` is when your post will be published.
# For the most part, you can leave this as the day you _started_ the post.
# The maintainers will update this value before publishing
# The time is generally irrelvant in how Valkey published, so '01:01:01' is a good placeholder
date= 2024-08-02 01:01:01
# 'description' is what is shown as a snippet/summary in various contexts.
# You can make this the first few lines of the post or (better) a hook for readers.
# Aim for 2 short sentences.
description= "Delivering significant improvements in throughput, scalability, and observability for modern data-intensive applications"
# 'authors' are the folks who wrote or contributed to the post.
# Each author corresponds to a biography file (more info later in this document)
authors= [ "pingxie", "madolson" ]
+++

> # Valkey 8.0: Delivering Enhanced Performance and Reliability

Reviewer: This automatically get's added at the top. 
```suggestion
```

> The Valkey community is proud to unveil the first release candidate of Valkey 8.0,
> Valkey 8.0 RC1, a major update designed to enhance performance, reliability,
> and observability for all Valkey installations.

Reviewer: ```suggestion
The Valkey community is proud to unveil the first release candidate of Valkey 8.0,
Valkey 8.0 RC1, a major update designed to enhance performance, reliability,
and observability for all Valkey installations. In this blog, we'll dive a bit deeper
into each of these areas and talk about the exciting features in this new release.
```

to a more responsive system, even during synchronization. See [#60](https://github.com/valkey-io/valkey/pull/60) for details.

## Observability

Valkey 8.0 introduces a comprehensive per-slot metrics infrastructure, providing
detailed visibility into the performance and resource usage of individual slots.
This granular data helps inform decisions about resource allocation, load
balancing, and performance optimization.

- **Key Count**: Returns the number of keys in each slot, making it easier to identify the slots with the largest number of keys.
- **CPU Usage**: Tracks CPU time consumed by operations on each slot, identifying
  areas of high utilization and potential bottlenecks.
- **Network Input/Output Bytes**: Monitors data transmission and reception by
  each slot, offering insights into network load and bandwidth utilization.
- **Minimal Overhead**: Initial benchmarks show that enabling detailed metrics
  incurs a negligible overhead of approximately 0.7% in QPS.

## Efficiency

Valkey 8.0 introduces two new improvements that reduce the memory overhead of keys, allowing users to store more data without any application changes.

> The change change is that keys are now embedded in the main dictionary, eliminating separate

Reviewer: ```suggestion
The first change is that keys are now embedded in the main dictionary, eliminating separate
```

+++
# `title` is how your post will be listed and what will appear at the top of the post
title= "Valkey 8.0: Delivering Enhanced Performance and Reliability"
# `date` is when your post will be published.
# For the most part, you can leave this as the day you _started_ the post.
# The maintainers will update this value before publishing
# The time is generally irrelvant in how Valkey published, so '01:01:01' is a good placeholder
date= 2024-08-02 01:01:01
# 'description' is what is shown as a snippet/summary in various contexts.
# You can make this the first few lines of the post or (better) a hook for readers.
# Aim for 2 short sentences.

> description= "Delivering significant improvements in throughput, scalability, and observability for modern data-intensive applications"

Reviewer 2: is this the new title? seems quite lengthy

Reviewer: No, I honestly don't know what this is. "Valkey 8.0: Delivering Enhanced Performance and Reliability"

> description= "Delivering significant improvements in throughput, scalability, and observability for modern data-intensive applications"

Reviewer: ```suggestion
description= "The first release candidate of Valkey 8.0 is now available! Come learn about the exciting improvements in performance, reliability, and observability that are available in this new version."
```

- **Intelligent Core Utilization**: Distributes I/O tasks across multiple
  cores based on realtime usage, reducing idle time and improving energy efficiency.
- **Command Batching**: Optimizes memory access patterns by prefetching frequently
  accessed data to minimize CPU cache misses, reducing memory accesses required for
  dictionary operations.

For more details on these improvements, you can refer to
[#758](https://github.com/valkey-io/valkey/pull/758) and
[#763](https://github.com/valkey-io/valkey/pull/763).

## Reliability

Cluster scaling operations via slot migrations have historically been delicate.
Valkey 8.0 improves reliability and minimizes disruptions with the following
enhancements:

- **Automatic Failover for Empty Shards**: New shards that start empty, owning
  no slots, now benefit from automatic failover. This ensures high availability
  from the start of the scaling process.
- **Replication of Slot Migration States**: All `CLUSTER SETSLOT` commands are
  now replicated synchronously to replicas before execution on the primary. This
  reduces the chance of unavailability if the primary fails, as the replicas have

>   the most up-to-date information about the state of the shard. New Replicas also

Reviewer 2: nit - should replicas be capitalized?

>   the most up-to-date information about the state of the shard. New Replicas also

Reviewer: ```suggestion
  the most up-to-date information about the state of the shard. New replicas also
```

  updates the slot migration states on source and target nodes. This ensures requests
  are continuously routed to the correct primary in the target shard, maintaining
  cluster integrity and availability.

For more details on these improvements, you can refer to
[#445](https://github.com/valkey-io/valkey/pull/445).

## Replication

Valkey 8.0 introduces a dual-channel replication scheme, allowing the RDB and
the replica backlog to be transferred simultaneously, accelerating synchronization.

- **Reduced Memory Load**: By streaming replication data to the replica during
  the full sync, the primary node experiences significantly less memory pressure.
  The replica now manages the Client Output Buffer (COB) tracking, reducing the
  likelihood of COB overruns and enabling larger COB sizes on the replica side.
- **Reduced Parent Process Load**: A dedicated connection for RDB transfer frees
  the primary's parent process from handling this data, allowing it to focus on
  client queries and improving overall responsiveness.

Performance tests show improvements in write latency during sync, and in scenarios

> with heavy read commands, the sync time can be cut by up to 50%. This translates

Reviewer 2: do we have source for this? "up to" doesn't make me as worried, but just figured I'd note it.

Reviewer: It's detailed in the issue linked below.

reducing the memory overhead by 16 bytes per key-value pair without degrading performance. 

For more details on these improvements, you can refer to [#541](https://github.com/valkey-io/valkey/pull/541)
and [Redis#11695](https://github.com/redis/redis/pull/11695).

## Additional Highlights

- **Dual IPv4 and IPv6 Stack Support**: Seamlessly operate in mixed IP environments
  for enhanced compatibility and flexibility.
  See [#736](https://github.com/valkey-io/valkey/pull/736) for details.
- **Improved Pub/Sub Efficiency**: Lightweight cluster messages streamline
  communication and reduce overhead for faster, more efficient Pub/Sub operations.
  See [#654](https://github.com/valkey-io/valkey/pull/654) for details.
- **Valkey Over RDMA (Experimental)**: Unlock significant performance improvements
  with direct memory access between clients and Valkey servers, delivering up to
  275% increase in throughput.
  See [#477](https://github.com/valkey-io/valkey/pull/477) for details.
- **Numerous Smaller Performance/Reliability Enhancements**: Many under-the-hood
  improvements ensure a smoother, more stable experience across the board.

>   See [release notes](https://github.com/valkey-io/valkey/releases/edit/8.0.0-rc1) for details.

Reviewer: ```suggestion
  See [release notes](https://github.com/valkey-io/valkey/releases/tag/8.0.0-rc1) for details.
```
Probably not the edit link.
